src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
  * 02111-1307, USA.
  *)
 
-(** Module [Lwt]: cooperative light-weight threads. *)
 
-(** This module defines {e cooperative light-weight threads} with
-    their primitives. A {e light-weight thread} represent a
-    computation that may be not terminated, for example because it is
-    waiting for some event to happen.
 
-    Lwt threads are cooperative in the sense that switching to another
-    thread is always explicit (with {!wakeup} or {!wakeup_exn}). When a
-    thread is running, it executes as much as possible, and then
-    returns (a value or an error) or sleeps.
+(** Asynchronous programming with promises.
 
-    Note that inside a Lwt thread, exceptions must be raised with
-    {!fail} instead of [raise]. Also the [try ... with ...]
-    construction will not catch Lwt errors. You must use {!catch}
-    instead. You can also use {!wrap} for functions that may raise
-    normal exception.
+    {b Promises} are placeholders for values that take a long time to compute.
+    Promises are similar to [ref]s – they can store one value. Here is how they
+    differ from [ref]s:
 
-    Lwt also provides the syntax extension {!Pa_lwt} to make code
-    using Lwt more readable.
-*)
+    - A promise might not have a value yet. This is called a {e pending}
+      promise.
+    - Writing a value into a promise is called {e resolving} it.
+    - It's possible to attach {b callbacks} to a pending promise. They will be
+      run when the promise gets resolved.
```

Comment:
There should be a mention early on that the callbacks typically resolve another promise. Not necessarily here, but maybe bellow just after as part of the answer in “why it is this useful?”

Reply:
See later comment for discussion. I think it's better to show that through examples (to be created later), rather than complicate the promise/callback idea. It is mentioned in execution model, that callbacks often resolve more promises.

I chose to focus on a single "step" of execution here, rather than the whole Lwt "loop." I hope it gives a simpler and more local view to the reader.